mod_bot/gym.py

```diff
@@ -0,0 +1,111 @@
+import Levenshtein
+
+candidates_key = "candidates"
+
+
+def get_approx_name(gym_name, gym_list):
+    candidates_list = []
+    try:
+        gn = clean_name(gym_name)
+
+        # Try autocomplete
+        keys = gym_list.keys()
+        gyms = list(map(lambda x: clean_name(x), keys))
+
+        candidates = [gym for gym in gyms if gn in gym]
+        candidates = list(map(lambda x: find_gym_from_clean_name(x, gyms, gym_list), candidates))
+        if len(candidates) == 1:
+            return candidates[0]
+
+        candidates_list.extend(candidates)
+
+        # Try short_name
+        short_gyms = list(map(lambda x: clean_name(x[0]), gym_list.values()))
+        candidates = [gym for gym in short_gyms if gn in gym]
+        candidates = list(map(lambda short_name: find_gym_from_short_name(short_name, gym_list), candidates))
+        if len(candidates) == 1:
+            return candidates[0]
+
+        candidates_list.extend(candidates)
+
+        # Try graph autocomplete
+        tree = build_full_name_graph(gym_list)
+        candidates = autocomplete_from_tree(tree, gn)
+        candidates = list(map(lambda x: find_gym_from_clean_name(x, gyms, gym_list), candidates))
+        if len(candidates) == 1:
+            return candidates[0]
+
+        candidates_list.extend(candidates)
+
+        # Try Levenshtein distance
+        for k, v in gym_list.items():
+            d = Levenshtein.distance(gn, k)
+            if d < 3:
+                return v
+    except Exception as e:
+        # Failed.
+        print(e)
+        pass
+    return set(candidates_list)
+
+
+def find_gym_from_short_name(short_name, gym_list):
+    return list(filter(lambda x: clean_name(x[0]) == short_name, gym_list.values()))[0]
+
+
+def find_gym_from_clean_name(clean_gn, gyms, gym_list):
```

Comment:
Wouldn't passing `gym_list.keys` be better as a parameter, or at least easier to read than `gym_list[list(gym_list.keys())[index]]` ?

Reply:
I must admit this is far from easy to read. I will decompose it